packages/utils/src/lib/scoring.ts

```diff
@@ -24,103 +24,88 @@ export type ScoredReport = Omit<Report, 'plugins' | 'categories'> & {
 ): number {
-  const numerator = refs.reduce(
-    (sum, ref) => sum + scoreFn(ref) * ref.weight,
-    0,
+  const { numerator, denominator } = refs.reduce(
+    (acc, ref) => {
+      const score = scoreFn(ref);
+      return {
+        numerator: acc.numerator + score * ref.weight,
+        denominator: acc.denominator + ref.weight,
+      };
+    },
+    { numerator: 0, denominator: 0 },
   );
-  const denominator = refs.reduce((sum, ref) => sum + ref.weight, 0);
   return numerator / denominator;
 }
 
 export function scoreReport(report: Report): ScoredReport {
-  const scoredPlugins = report.plugins.map(plugin => {
-    const { groups, audits } = plugin;
-    const preparedAudits = audits.map(audit => ({
-      ...audit,
-      plugin: plugin.slug,
-    }));
-    const preparedGroups =
-      groups?.map(group => ({
-        ...group,
-        score: calculateScore(group.refs, groupRefToScore(preparedAudits)),
-        plugin: plugin.slug,
-      })) || [];
+  const allScoredAuditsAndGroups = new Map();
 
-    return {
-      ...plugin,
-      audits: preparedAudits,
-      groups: preparedGroups,
-    };
+  report.plugins.forEach(plugin => {
+    const { audits } = plugin;
+    const groups = plugin.groups || [];
+
+    audits.forEach(audit => {
+      const key = `${plugin.slug}-${audit.slug}-audit`;
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      audit.plugin = plugin.slug;
+      allScoredAuditsAndGroups.set(key, audit);
+    });
+
+    function groupScoreFn(ref: AuditGroupRef) {
+      const score = allScoredAuditsAndGroups.get(
+        `${plugin.slug}-${ref.slug}-audit`,
+      )?.score;
+      if (score == null) {
+        throw new Error(
+          `Group has invalid ref - audit with slug ${plugin.slug}-${ref.slug}-audit not found`,
+        );
+      }
+      return score;
+    }
+
+    groups.forEach(group => {
+      const key = `${plugin.slug}-${group.slug}-group`;
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      group.score = calculateScore(group.refs, groupScoreFn);
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-expect-error
+      group.plugin = plugin.slug;
+      allScoredAuditsAndGroups.set(key, group);
+    });
+    plugin.groups = groups;
   });
 
-  // @TODO intro dict to avoid multiple find calls in the scoreFn
-  const allScoredAudits = scoredPlugins.flatMap(({ audits }) => audits);
-  const allScoredGroups = scoredPlugins.flatMap(({ groups }) => groups);
+  function catScoreFn(ref: CategoryRef) {
+    const key = `${ref.plugin}-${ref.slug}-${ref.type}`;
+    const item = allScoredAuditsAndGroups.get(key);
+    if (!item) {
+      throw new Error(
+        `Category has invalid ref - ${ref.type} with slug ${key} not found in ${ref.plugin} plugin`,
+      );
+    }
+    return item.score;
+  }
+
+  const scoredCategoriesMap = report.categories.reduce(
```

Comment:
I agree, even though `reduce` is nicer, more readable, no-sideffects, ... 😄 

We can just stick to `for` (not even `forEach`) plain `for`, iterate fast and optimized inside and have a nice functional API.

why not even `forEach`? because it creates copy of each of the iterated items for passing as parameter (same as `map`, `filter`, `reduce`), and that is unnecessary overhead.

Reply:
Changed to for...of, performance is improved